### PR TITLE
Add TensorflowHub embeddings to MLTransform

### PR DESCRIPTION
**Pull request by AnandInguva**
TODO: https://github.com/apache/beam/issues/30288, also mentioned in the code.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            | "MLTransform" >> MLTransform(
+                write_artifact_location=self.artifact_location).with_transform(
+                    embedding_config))
+
+
+@unittest.skipIf(tf is None, 'Tensorflow is not installed.')
+class TFHubEmbeddingsGCSArtifactLocationTest(TFHubEmbeddingsTest):
+  def setUp(self):
+    self.artifact_location = os.path.join(
+        'gs://temp-storage-for-perf-tests/vertex_ai', uuid.uuid4().hex)
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py:166]:
```suggestion
        'gs://temp-storage-for-perf-tests/tfhub', uuid.uuid4().hex)
```

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+try:
+  import tensorflow as tf  # disable=unused-import
+  from apache_beam.ml.transforms.embeddings.tensorflow_hub import TensorflowHubTextEmbeddings
+except ImportError:
+  tf = None
+
+try:
+  from apache_beam.ml.transforms.tft import ScaleTo01
+except ImportError:
+  ScaleTo01 = None  # type: ignore
```
Comment (riteshghorse) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py:40]: (Optional) Nit: may narrow the scope to `ignore[assignment]`

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    Args:
+      columns: The columns containing the text to be embedded.
+      hub_url: The url of the tensorflow hub model.
+      preprocessing_url: The url of the preprocessing model. This is optional.
+        If provided, the preprocessing model will be used to preprocess the
+        text before feeding it to the main model.
+      min_batch_size: The minimum batch size to be used for inference.
+      max_batch_size: The maximum batch size to be used for inference.
+      large_model: Whether to share the model across processes.
+    """
```
Comment (riteshghorse) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py:109]: Should we specify that these args should be passed with `kwargs`?

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    # override the default inference function
+    return _TensorflowHubModelHandler(
+        model_uri=self.model_uri,
+        preprocessing_url=self.preprocessing_url,
+        min_batch_size=self.min_batch_size,
+        max_batch_size=self.max_batch_size,
+        large_model=self.large_model,
+    )
+
+  def get_ptransform_for_processing(self, **kwargs) -> beam.PTransform:
```
Comment (riteshghorse) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py:124]: Nit: consider adding a doc comment

**Review by riteshghorse — APPROVED** [review]
Minor nits, but LGTM!

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    PTransform with the help of a ModelHandler.
+    Args:
+      columns: The columns containing the text to be embedded.
+      hub_url: The url of the tensorflow hub model.
+      preprocessing_url: The url of the preprocessing model. This is optional.
+        If provided, the preprocessing model will be used to preprocess the
+        text before feeding it to the main model.
+      min_batch_size: The minimum batch size to be used for inference.
+      max_batch_size: The maximum batch size to be used for inference.
+      large_model: Whether to share the model across processes.
```
Comment (riteshghorse) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py:108]: are these part of kwargs or separate args?

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub.py:108]: part of **kwargs and args of EmbeddingManager. 

Added docstrings for informational purposes.

File: sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+try:
+  import tensorflow as tf  # disable=unused-import
+  from apache_beam.ml.transforms.embeddings.tensorflow_hub import TensorflowHubTextEmbeddings
+except ImportError:
+  tf = None
+
+try:
+  from apache_beam.ml.transforms.tft import ScaleTo01
+except ImportError:
+  ScaleTo01 = None  # type: ignore[assignment]
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/transforms/embeddings/tensorflow_hub_test.py:40]:
```suggestion
  ScaleTo01 = None  # type: ignore[misc]
```

**Comment by AnandInguva** [comment]
Merging this. Thanks.

**Pull request merged**